File: app/src/main/java/com/cbr/behancesampleapp/ui/common/mvp/BaseMvpFragment.java
package com.cbr.behancesampleapp.ui.common.mvp;

import android.app.Fragment;

import javax.inject.Inject;

/**
 * Created by Dimitrios on 8/26/2017.
 */

public abstract class BaseMvpFragment<PRESENTER extends MvpPresenter> extends Fragment implements MvpView {

	@Inject
	PRESENTER mPresenter;

	public PRESENTER getPresenter() {
		if (mPresenter == null) {
			throw new RuntimeException("You forgot to inject or set your Presenter :(");
		}
		return mPresenter;
	}

	public void setPresenter(PRESENTER presenter) {
		this.mPresenter = presenter;
	}


	@Override
	public void onResume() {
		super.onResume();
		getPresenter().subscribe();
	}

	@Override
	public void onPause() {
		super.onPause();
		getPresenter().unsubscribe();
	}

}

File: app/src/main/java/com/cbr/behancesampleapp/ui/common/mvp/BaseMvpFragment.kt
package com.cbr.behancesampleapp.ui.common.mvp

import android.os.Bundle
import android.support.v4.app.Fragment
import com.cbr.behancesampleapp.BehanceSampleApp

/** Created by Dimitrios on 8/26/2017. */
abstract class BaseMvpFragment : Fragment() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        onFragmentInject()
    }

    protected abstract fun onFragmentInject()

    fun appComponent() = BehanceSampleApp.appComponent
}
